Move opkssh key files to a separate location

net42-jkeil opened the pull request:
This PR changes how the key files opkssh creates on the client are stored.

Addressed problems:
It prevents problems in case the predefined allowed identity files (`id_ecdsa` / `id_ed25519`) are already in use. (#69)
If a user has to log into multiple systems handled by different providers the user has to run `opkssh login` and `ssh` in that order for each system because repeated calls of `opkssh login` will override the PK of the previous provider.

Changes:

- Identity files opkssh creates are named unique per provider and clientID
- Additionally they get created in `~/.ssh/opkssh/` to reduce clutter and make cleaning up easier
- Identity management happens in `~/.ssh/opkssh/config`

Notable:
Identities that do not follow the standard naming convention or are not placed in the default location `~/.ssh/` are usually ignored. To fix this the `Include ~/.ssh/opkssh/config` directive has to be added to the users ssh config.
This is to prevent possible corruption when regularly modifying the users ssh config.

The README has been modified to reflect this change. Additionally the command `opkssh config` has been added which checks for correct configuration and instructs the user how to modify their config if needed.

With the upcoming PR #96 most users will not encounter those problems, but on systems without the ssh-agent this remains a problem.

MrCriseas commented:
I know, we are from the same organisation but would be nice. Not everyone is using ssh-agent and the package would be directly useable without setting up the agent. Nevertheless to potentially overwrite an ssh certificate and check is it used beforehand by opk seems unlikely not an optimal solution.

We have the usecase to use multiple providers and that solution could address our needs massively and support this repository :).

EthanHeilman commented:
Want to talk about this at this months community meeting?

OpenPubkey Community Meeting 2:00 PM (ET) 6:00pm (GMT)
Wednesday, April 16 · 2:00 – 3:00pm
Time zone: America/New_York
Google Meet joining info
Video call link: https://meet.google.com/hvc-dywm-wzk

net42-jkeil commented:
As mentioned in the community meeting this PR will be on hold until client configuration is implemented.
This PR might need a complete rewrite at that point to cleanly integrate with the new configuration method.

EthanHeilman commented:
Merged the client config.

We don't automatically create the client config (maybe we should).
https://github.com/openpubkey/opkssh/pull/143

net42-jkeil commented:
Finally got around to refactor the PR to use the client config.

The functionality changed slightly. I now use the `~/.opk` directory for the opk ssh config. This seemed like the right location to me.

The opk config.yml got extended to include `key_management` with a `default_key_dir` and `use_identity_config` directive.
Those set the directory for opk's keys and whether or not it should add those keys as `IdentityFile` entries to a ssh config in `~/.opk/config`.
If true, it adds the `Include` directive to the users ssh config.

I added this information to the readme.

Users with an existing config and users that create the config after this PR will see no change to the default behavior. It requires manual changes to the config to take effect.

EthanHeilman commented:
@net42-jkeil The client config is only used by login at the moment so it seems natural that  login manage it. That said, I'm not sure exactly what you mean here. What would the SetConfig function do? What is the intent?

net42-jkeil commented:
With my PR the `login` function checks `l.config.KeyManagement.DefaultKeyDir` for a modified config.
If the variable is set it will handle the key placement and if nessesary add an `Identity`.

Alternatively I could extend the if check to prevent it from accessing the uninitialized `config` variable.
I was not expecting `config` to be uninitialized in the integration testing since it is always set in normal execution.

EthanHeilman commented:
When you say `config` do you mean the client config `~/.opk/config.yml`? I was originally opposed to automatically creating `~/.opk/config.yml` on first call to `opkssh login` but I've been rethinking that. If it is easier to assume `~/.opk/config.yml` exists, then create it on first `opkssh login` call.

net42-jkeil commented:
What I meant was a variable in code.
I made an assumption in code that was not always true. This assumption was corrected in the last commit.

This way I could remove my changes to the integration test, too.

I for my part do not require a `~/.opk/config.yml`.

EthanHeilman commented:
Sorry, I misunderstood you.

The [config variable on the login command struct ](https://github.com/openpubkey/opkssh/blob/016409ca011d74c695f2007cc10fe97be6337a84/commands/login.go#L65)should always be set regardless of it if a config file exists or not. It is set in `login.Run`.

But yeah`l.config.KeyManagement.DefaultKeyDir` might not be set.

net42-jkeil commented:
@EthanHeilman let me know once you got to check the PR and if there is anything preventing a merge.
Thanks!

EthanHeilman commented:
I'll have more free time on Thursday, so I don't get a review done by Wednesday, I'll give it a review Thursday night.

EthanHeilman commented on commands/login.go:
```diff
@@ -497,18 +513,161 @@ func (l *LoginCmd) writeKeysToSSHDir(seckeySshPem []byte, certBytes []byte) erro
+	)
+
+	// build user SSH dir
+	userHomeDir, err = os.UserHomeDir()
```
```suggestion
	userHomeDir, err := os.UserHomeDir()
```

EthanHeilman commented on test/integration/login_test.go:
```diff
@@ -27,11 +27,12 @@ import (
 	"github.com/stretchr/testify/require"
 	"golang.org/x/crypto/ssh"
+
+	"github.com/openpubkey/opkssh/commands"
```
Revert this change to keep git history clean

EthanHeilman commented on commands/config/client_config.go:
```diff
@@ -42,3 +45,34 @@ func NewClientConfig(c []byte) (*ClientConfig, error) {
+
+	// if default key dir is not set we use the default logic
+	if c.KeyManagement.DefaultKeyDir == "" {
+		return
```
I recognize this syntax of declaring variables in the output block and then letting return has some big advantages, but I don't want to use it in opkssh because it makes it harder to understand what is happening.

EthanHeilman commented on commands/login.go:
```diff
@@ -497,18 +513,161 @@ func (l *LoginCmd) writeKeysToSSHDir(seckeySshPem []byte, certBytes []byte) erro
+	issuer, _ = strings.CutPrefix(issuer, "https://")
+	issuer = regex.ReplaceAllString(issuer, "_")
+
+	return issuer
```
These issue filenames are going to be huge for things like azure.

net42-jkeil replied:
Correct, would 40 characters be a reasonable length? This is what tells the user what provider the key is from and should include the realm to make it unique.

EthanHeilman replied:
I'd do provider_alias+"-"+IDToken.email+"-"+hex(SHA256(issuer,client-id))[0:7])

so `google-alice@gmail.com-67ea488.pub`

If no email exists in the ssh-key then provider_alias+"-"+IDToken.sub+"-"+hex(SHA256(client-id)[0:4]) subs can annoyingly long, but most ID Tokens have email

`google-AAAAAAAAAAAAAAAAAAAAAJ8PFm0pjpXKQouYRalE11g-67ea488.pub`

We can store the full client-id in the SSH public key comment. In the example below the client-id is `bd345b9c-6902-400d-9e18-45abdf0f698f`

```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAAgQDSNM6RVVmwN3y0NurIQnmZgjcx5K5zzZu9nDqopW4J
In/mr8OYZI3heSJShnIM8EThvwVGXXXyyJVRQAvRHYFO4DxS6bufSNWr3BxBGaGYlYxI9mgvQnT6+MzE
3oZyEMdQNPlV5VfbileXlrPoAk1TkGdVdhwdLJMI2B4KUyMf+Q== openpubkey:bd345b9c-6902-400d-9e18-45abdf0f698f
```

Thus, when writing ssh key we can check if the filename collides through random chance by checking the comment on the key we are planning to overwrite or if the client-ids are the same and we should overwrite. 

```
hex(SHA256(issuer1,client-id1))[0:7]) == hex(SHA256(issuer1,CLIEND-ID-2))[0:7])
```

The probability of a random collision if the same issuer and email is used for two different client-ids is 1/256 so it can happen but should be rare.

If we detect a collision, rather than overwriting we just extend the SHA256 hash length in filename until there isn't a collision.

`google-alice@gmail.com-67ea488.pub` (client-id-1) collides with `google-alice@gmail.com-67ea488.pub` (client-id-2)
so we extend `google-alice@gmail.com-67ea488.pub` --> `google-alice@gmail.com-67ea488f.pub`.

Lets say we have three SSH keys resulting from pass collisions and now we want to write a new ssh key for alice@gmail.com with client-id-4.

```
google-alice@gmail.com-67ea488.pub  # (client-id-1)
google-alice@gmail.com-67ea488f.pub  # (client-id-2)
google-alice@gmail.com-67ea488fe.pub #  (client-id-3)
```

Our new public key collides with `google-alice@gmail.com-67ea488.pub` so we length it to `google-alice@gmail.com-67ea488f.pub` but it collides again, so we length it again `google-alice@gmail.com-67ea488fe.pub` but it collides again so we length it one more time to `google-alice@gmail.com-67ea488fe4.pub` no collision. This should be rare as the probability of collisions decreases exponentially with each increase in length and the birthday paradox is unlikely to be come into effect here.

```
google-alice@gmail.com-67ea488.pub  # (client-id-1)
google-alice@gmail.com-67ea488f.pub  # (client-id-2)
google-alice@gmail.com-67ea488fe.pub #  (client-id-3)
google-alice@gmail.com-67ea488fe4.pub #  (client-id-4)
```

One thing to watch out for is a SSH server will only check `MaxAuthTries` public keys in a session before the SSH server automatically disconnects. If you have 7 keys and the 7th key is the correct one, SSH will disconnect you before trying that one.

[MaxAuthTries](https://man.openbsd.org/sshd_config#MaxAuthTries)
Specifies the maximum number of authentication attempts permitted per connection. Once the number of failures reaches half this value, additional failures are logged. The default is 6.
<https://man.openbsd.org/sshd_config#MaxAuthTries>

Some psychopaths set this to `MaxAuthTries=2`, but it is fairly rare and thats on them. We should assume `MaxAuthTries=6`.  If we ever find more than 6 keys for a particular provider-issuer pair, we should warn the user.

net42-jkeil replied:
After tinkering a bit I found that I could use the pkToken object to extract what I needed.
This seems to allow combinations like `issuer-audience` which represents `provider-clientID` just as originally planned:
```
issuer: auth.xyz.tld/realms/master
audience: opkssh

auth.xyz.io_realms_master_2336411d29ee5df78554739aa470430f18debb16336508be8382a5d23b56fb8d
```

We also have access to the `subscriber` field of the token. This is the unique identification of the user on the provider. For KeyCloak it is a GUID.

Adding this to the key name would allow multiple logins with the same provider clientID combination. This seems like a bad idea to me. There should only ever be one active login per provider and clientID otherwise you could login as either `jon@company.tld` OR `doe@company.tld`.

A case in favor for this are admins. Those could login as their normal user or a separate admin account that has access on more systems.

The JWT token can be parsed from the pkToken and provide access to all private claims.
For KeyCloak the EMail is accessible too, but availability for all providers and configurations is not guaranteed.

net42-jkeil replied:
> `MaxAuthTries`

The ssh client will offer the identity keys in order they were written in the config (FIFO).
So adding the new keys at the top will ensure that the most recently acquired key pair is used first.
If that is also the case if there were other keys loaded into the agent before I can not tell right now.

SSH does not know which keys to offer, so it will always offer all. Parsing the configs is also out of scope so we should rather check the amount of providers added to the config and assume there is one key per provider.

I'll add a warning for that in the README and log output.

EthanHeilman reviewed (changes requested):
Thanks for your hard work on this

net42-jkeil commented on commands/login.go:
```diff
@@ -302,10 +312,16 @@ func (l *LoginCmd) login(ctx context.Context, provider providers.OpenIdProvider,
 		if err := l.writeKeys(seckeyPath, seckeyPath+".pub", seckeySshPem, certBytes); err != nil {
 			return nil, fmt.Errorf("failed to write SSH keys to filesystem: %w", err)
 		}
+	} else if l.config != nil && l.config.KeyManagement.DefaultKeyDir != "" {
```
`KeyManagement` is not a reference so it will be initialized via the config object.

net42-jkeil commented on commands/config/default-client-config.yml:
```diff
@@ -1,6 +1,10 @@
 
+key_management:
+  default_key_dir:
+  use_identity_config: false
```
Setting default values will cause changes for existing users. I tried to prevent that.
Enabling use_identity_config will modify a users `.ssh/config` file. It is noted in the README, but again I wanted to prevent effects on users.

If you are ok with this we can set defaults as `~/.opk` and `true`.

EthanHeilman replied:
Please double check my work, but I believe changing the default config only impacts a user if they run `.\opkssh.exe login --create-config`. This command will fail if they already have a config.

Could we set `default_key_dir: ~/.ssh` and then only activate the new naming behavior if `use_identity_config: true` What would the UX be like on this?

> Enabling use_identity_config will modify a users .ssh/config file. It is noted in the README, but again I wanted to prevent effects on users.

I agree we shouldn't touch the user's .ssh/config file without asking them first. This should be no.

net42-jkeil replied:
> I believe changing the default config only impacts a user if they run  `.\opkssh.exe login --create-config`

If no config file was found and `--create-config` was not passed it will use the embedded default config. So there could be an impact depending on the changes.

> Could we set `default_key_dir: ~/.ssh` and then only activate the new naming behavior if `use_identity_config: true`

Yes that is possible.

Here are a few things ragarding UX with changed `default_key_dir` and `use_identity_config: false`:
- it would create an id_ecdsa or id_ed25519 (if non opkssh ecdsa exists) key in the specified location
- the ssh client will not auto-load those into the ssh-agent or for authentication
- repeated `login` calls will override the key
- requires `-i <key>` flag to use (path gets output after `login` call)

I'll set the default dir as it would chance nothing from the current logic.

net42-jkeil commented on commands/login.go:
```diff
@@ -497,18 +513,161 @@ func (l *LoginCmd) writeKeysToSSHDir(seckeySshPem []byte, certBytes []byte) erro
+
+		identityFileString = `IdentityFile ` + keyFileName
+
+		afs = &afero.Afero{Fs: l.Fs}
```
Thats just a my habit/style, I'll switch it to inline declaration.

net42-jkeil commented:
The git history should be corrected. README, error logs, var blocks and inline declaration changes made as requested.
I couldn't apply your suggestions because I had to rebase to fix the history.

EthanHeilman commented:
That was a fast turn around! No worry about applying my suggestions.  Suggestions typically break the linter because github doesn't apply lint suggests and my suggestions are intended to communicate an idea not be 100% correct code.

net42-jkeil commented on commands/config/key-management.go:
```diff
@@ -0,0 +1,42 @@
+func (k *KeyManagementConfig) GetKeyDir() (configKeyDir string, err error) {
+
+	if !strings.HasPrefix(k.DefaultKeyDir, "/") {
+		configKeyDir, err = os.UserHomeDir()
```
> Then if `default_key_dir: ~/.ssh/opkssh` the command `opkssh login -i opkssh_server_group1` would find the key `~/.ssh/opkssh/opkssh_server_group1`

I'll have to check that. If it does not work that way right now, I'll make it so it does unless there is a absolute path given.

> Would doing what I propose cause problems

No, the opk ssh config is hard coded to be created as `~/.opk/config` and included in `~/.ssh/config` to prevent exactly those problems.

EthanHeilman replied:
> No, the opk ssh config is hard coded to be created as ~/.opk/config and included in ~/.ssh/config to prevent exactly those problems.

What about naming it `opkssh_ids` so that it doesn't have to live in the `~/.opk` directory. It seems likely to cause confusion if we have `~/.opk/config.yml` (the client config) and `~/.opk/config` (SSH key identities). Then if the default directory is `~/.ssh` it wouldn't conflict with the ssh config.

net42-jkeil commented on commands/login.go:
```diff
@@ -497,18 +513,161 @@ func (l *LoginCmd) writeKeysToSSHDir(seckeySshPem []byte, certBytes []byte) erro
+	}
+
+	// get file name from issuer
+	keyFileName = makeSSHKeyFileName(issuer)
```
I found a way to access the needed info that I overlooked before and implemented as intended in the beginning.
Does this resolve this CR of yours?

net42-jkeil commented:
@EthanHeilman I think I addressed all suggestions and notes, please review when you have the time

EthanHeilman commented:
@net42-jkeil Going to take this for a test drive today and give it a review.

EthanHeilman commented:
I tried this, but I couldn't login in because I already had a config.yml which didn't define the `default_key_dir` so it defaulted to `~/` which meant ssh couldn't find it.


I added this to my config.yml
```
key_management:
  default_key_dir: ~/.ssh
  use_identity_config: false
```

but I got the following error

```
2025/05/22 16:29:57 Error executing login command: error logging in: failed to write SSH keys to ssh dir: open C:\Users\e0\.ssh\C:\Users\e0\.ssh\id_ecdsa: The filename, directory name, or volume label syntax is incorrect.
Error: error logging in: failed to write SSH keys to ssh dir: open C:\Users\e0\.ssh\C:\Users\e0\.ssh\id_ecdsa: The filename, directory name, or volume label syntax is incorrect.
```

net42-jkeil commented:
@EthanHeilman I completely missed Windows here. I have no environment to test it, but using filepath build-ins should do the trick.
Please let me know if that fixed the problems :+1:

EthanHeilman commented:
@net42-jkeil Checked and it works on windows now when

```
key_management:
  default_key_dir: ~/.ssh
  use_identity_config: false
```
or no key_management set

**1. Testing with `default_key_dir != ~/.ssh` and `use_identity_config = false`**

When I change this to

```
key_management:
  default_key_dir: ~/.opk/config/
  use_identity_config: false
```

I get

```
PS C:\Users\e0\Documents\GitHub\opkssh> .\opkssh.exe login
2025/05/25 14:26:57 Error executing login command: failed key dir check: could not get stats of key directory: CreateFile C:\Users\e0\.opk\config: The system cannot find the file specified.
Error: failed key dir check: could not get stats of key directory: CreateFile C:\Users\e0\.opk\config: The system cannot find the file specified.
```

After creating this directory it worked.

This isn't a bug, but I think it is a better UX if we automatically create the directory for the user. When ` default_key_dir != ~/.ssh` and `use_identity_config != false` we should output a warning to the console that the ssh might not be able to find the ssh keys.

**2 . Testing with ` default_key_dir = ~/.opk/config/` and `use_identity_config == true`**

I ran into the issue that if not ssh config exists I get

```
025/05/25 14:37:19 Error executing login command: error logging in: failed to write SSH keys to configured key dir (~/.opk/config/): failed to read opk ssh config: read C:\Users\e0\.opk\config: Incorrect function.
Error: error logging in: failed to write SSH keys to configured key dir (~/.opk/config/): failed to read opk ssh config: read C:\Users\e0\.opk\config: Incorrect function.
```

It did create this file `C:\Users\e0\.ssh\config.new` but ssh didn't seem to find it.

2025/05/25 14:45:12 Error executing login command: error logging in: failed to write SSH keys to configured key dir (~/.opk/config/): failed to read opk ssh config: read C:\Users\e0\.opk\config: Incorrect function.
Error: error logging in: failed to write SSH keys to configured key dir (~/.opk/config/): failed to read opk ssh config: read C:\Users\e0\.opk\config: Incorrect function.

**3 . Testing with ` default_key_dir = ~/.ssh/opkssh/` and `use_identity_config == true`**

Created `~/.ssh/opkssh/`,
Created `~/.ssh/config`,
Created `~/.ssh/opkssh/config`

```
Writing opk ssh public key to C:\Users\e0\.ssh\opkssh\accounts.google.com-206584157355-7cbe4s6.pub and corresponding secret key to C:\Users\e0\.ssh\opkssh\accounts.google.com-206584157355-7cbe4s6
2025/05/25 14:54:52 Error executing login command: error logging in: failed to write SSH keys to configured key dir (~/.ssh/opkssh/): failed to read opk ssh config: read C:\Users\e0\.opk\config: Incorrect function.
Error: error logging in: failed to write SSH keys to configured key dir (~/.ssh/opkssh/): failed to read opk ssh config: read C:\Users\e0\.opk\config: Incorrect function.
```

It creates a file `~/.ssh/config.new` whose contain is:
```
Include C:\Users\e0\.opk\config
```

I changed this to the correct value `Include C:\Users\e0\.ssh\opkssh\config` and reran `opkssh login` it changed it to.

```
Include C:\Users\e0\.opk\config
Include C:\Users\e0\.ssh\opkssh\config
```

So I can't get it to work

net42-jkeil commented:
Thank you for your review!

All of the `Incorrect function` errors were caused by using the config file paths as paths for directories. Once Go tries to access them as files this is the error you get.
As you noticed at the end those errors were corrected after removing the directories.
I implemented a simple check to warn the user if they do something similar.

> better UX if we automatically create the directory for the user

I implemented this so that the config directory `~/.ssh` also gets created if `use_identity_config = true`.
Path creation to the desired key dir was corrected.

> It did create this file `C:\Users\e0\.ssh\config.new` but ssh didn't seem to find it.

This is only a temporary copy of the users SSH config I create. It is modified to include opkssh's SSH config and renamed to `~/.ssh/config` overriding the original once I am sure that all steps were successful.

Testing in a Windows 10 VM things seem to work for me. Please retest with other paths.
```
default_key_dir: ~/.opk/mykeys
```

EthanHeilman commented:
I'm been giving this PR a lot of thought and I came to the realization that we are using the config file option `use_identity_config` to flag to opkssh to perform a setup task that we can't undo. Namely writing to `~/.ssh/config`. We can't undo it because we can't tell the lines we add from the lines the user adds if the key directory changes. It is like we are trying to control the ssh config in the opkssh config.

This is my proposal:

We get rid of `use_identity_config` and `default_key_directory` and instead add `opkssh login --configure` which performs the following actions:

1. Creates the directory `~/.ssh/opkssh/`, 
2. Creates the file `~/.ssh/opkssh/keys`, 
3. Appends the line `Include ~/.ssh/opkssh/keys` to `~/.ssh/config` where the `keys` file does the current job of `~/.opk/config` but renamed to avoid confusion with `~/.opk/config.yml`.

To determine if we should write to `~/.ssh/id_ecdsa` or use new key names we check if the file `~/.ssh/opkssh/keys` exists or not.

- This reduces the chance someone breaks opkssh by editing the `default_key_directory` config option.
- We don't have to handle the `use_identity_config` `true` --> `false` transition.

What do you think about this?

EthanHeilman commented:
> However naming the configuration file keys is misleading. I would rather call it config too, since it is part of the users SSH config once included.

Agreed, this is less likely to be confusing if it is in the .ssh direcotry

EthanHeilman commented:
I don't have time this week, but I'm going to make a push to get this reviewed next week

EthanHeilman reviewed (approved):
LGTM

EthanHeilman commented:
@net42-jkeil Thanks for all your hard work on this. Really appreciate it and sticking with the PR review. This was an usually long review process. The benefit this brings to project especially when a user has different IDPs for different servers.
